src/fees/compute_fees.py

```diff
@@ -0,0 +1,606 @@
+from abc import ABC, abstractmethod
+from copy import deepcopy
+from dataclasses import dataclass
+from fractions import Fraction
+import math
+import os
+from typing import Any
+
+
+from dotenv import load_dotenv
+from eth_typing import Address
+from hexbytes import HexBytes
+from web3 import Web3
+from web3.logs import DISCARD
+from web3.types import TxData, TxReceipt, EventData
+
+from src.constants import (
+    SETTLEMENT_CONTRACT_ADDRESS,
+    REQUEST_TIMEOUT,
+)
+from contracts.gpv2_settlement_abi import gpv2_settlement_abi
+
+import requests
+
+# types for trades
+
+
+@dataclass
+class Trade:
+    """Class for"""
+
+    order_uid: HexBytes
+    sell_amount: int
+    buy_amount: int
+    owner: HexBytes
+    sell_token: HexBytes
+    buy_token: HexBytes
+    limit_sell_amount: int
+    limit_buy_amount: int
+    kind: str
+    sell_token_clearing_price: int
+    buy_token_clearing_price: int
+
+    def volume(self) -> int:
```

Comment:
@fhenneke is the naming of the function accurate? In case there is a protocol fee, this is not really the volume but rather the remaining amount after the protocol fee has been subtracted

Reply:
That is the definition of volume I use. The same as for surplus. It is not the hypothetical "what would volume have been if there were no fee?" but rather "what is the volume?".

We could have another `raw_volume` function but that would not be used for anything. Or is there a use case?

Reply:
I guess one issue i have is that if we talk about, e.g., an order having 1% volume fee, this volume we are talking about here is not the one we should check in order to compute that fee.

Reply:
This is exactly the same issue as for surplus, is it not?

Because of that we always have to use modified factors in computing protocol fees in the autopilot, e.g. [here](https://github.com/cowprotocol/solver-rewards/blob/b7e74aea02f104122e03bd1b723ade735e5bacb1/queries/orderbook/prod_batch_rewards.sql#L177-L182).

Reply:
I would argue that surplus is mainly from a user point of view so defining surplus for a sell order as executed_buy_amount - limit_price makes sense to me. I.e., since we explicitly say that score is surplus plus protocol fees, referring to surplus as the portion the user receives seems fine.

But i haven't thought too much about it, tbh.

Reply:
To me it is perfectly consistent to use this very function for computing volume. This _is_ the volume (in the surplus token) of the order. Dune uses that number for volume as well in some cases (if there is no price for the other token).

We need exactly this function to compute protocol fees. In principle it could be renamed. But I am in favor of keeping the name.

Using a different definition is not possible _unless_ a volume factor is added as parameter
```python
def volume_before_fees(self, volume_factor):
    if self.kind == "sell":
        return self.buy_amount / (1 - volume_factor)
    if self.kind == "buy":
        return self.sell_amount / (1 + volume_factor)
```
So the volume would then depend on which fees are applied. It would not be a property of the trade itself. Might be a helpful function.